openapi2kong/openapi2kong.go

```diff
@@ -53,263 +56,209 @@ func (opts *O2kOptions) setDefaults() {
-		}
-		return name, nil
+func getKongName(extensions *orderedmap.Map[string, *yaml.Node]) (string, error) {
+	if extensions == nil {
+		return "", nil
 	}
-	return "", nil
-}
-
-func dereferenceJSONObject(
-	value map[string]interface{},
-	components *map[string]interface{},
-) (map[string]interface{}, error) {
-	var pointer string
 
-	switch value["$ref"].(type) {
-	case nil: // it is not a reference, so return the object
-		return value, nil
+	xKongName, ok := extensions.Get("x-kong-name")
 
-	case string: // it is a json pointer
-		pointer = value["$ref"].(string)
-		if !strings.HasPrefix(pointer, "#/components/x-kong/") {
-			return nil, fmt.Errorf("all 'x-kong-...' references must be at '#/components/x-kong/...'")
-		}
-
-	default: // bad pointer
-		return nil, fmt.Errorf("expected '$ref' pointer to be a string")
+	if !ok {
+		return "", nil
 	}
 
-	// walk the tree to find the reference
-	segments := strings.Split(pointer, "/")
-	path := "#/components/x-kong"
-	result := components
-
-	for i := 3; i < len(segments); i++ {
-		segment := segments[i]
-		path = path + "/" + segment
-
-		switch (*result)[segment].(type) {
-		case nil:
-			return nil, fmt.Errorf("reference '%s' not found", pointer)
-		case map[string]interface{}:
-			target := (*result)[segment].(map[string]interface{})
-			result = &target
-		default:
-			return nil, fmt.Errorf("expected '%s' to be a JSON object", path)
-		}
+	var name string
+	err := yaml.Unmarshal([]byte(xKongName.Value), &name)
+	if err != nil {
+		return "", fmt.Errorf("expected 'x-kong-name' to be a string: %w", err)
 	}
 
-	return *result, nil
+	return name, nil
 }
 
 // getXKongObject returns specified 'key' from the extension properties if available.
 // returns nil if it wasn't found, an error if it wasn't an object or couldn't be
 // dereferenced. The returned object will be json encoded again.
-func getXKongObject(props openapi3.ExtensionProps, key string, components *map[string]interface{}) ([]byte, error) {
-	if props.Extensions != nil && props.Extensions[key] != nil {
-		var jsonBlob interface{}
-		_ = json.Unmarshal(props.Extensions[key].(json.RawMessage), &jsonBlob)
-		jsonObject, err := jsonbasics.ToObject(jsonBlob)
-		if err != nil {
-			return nil, fmt.Errorf("expected '%s' to be a JSON object", key)
-		}
+func getXKongObject(
+	extensions *orderedmap.Map[string, *yaml.Node],
+	key string, components *map[string]interface{},
+) ([]byte, error) {
+	if extensions == nil {
+		return nil, nil
+	}
 
-		object, err := dereferenceJSONObject(jsonObject, components)
-		if err != nil {
-			return nil, err
-		}
-		return json.Marshal(object)
+	xKongObject, ok := extensions.Get(key)
+	if !ok || xKongObject == nil {
+		return nil, nil
 	}
-	return nil, nil
+
+	xKongObjectBytes, err := convertYamlNodeToBytes(xKongObject)
+	if err != nil {
+		return nil, fmt.Errorf("expected '%s' to be a YAML object", key)
+	}
+
+	var jsonBlob interface{}
+	_ = yaml.Unmarshal(xKongObjectBytes, &jsonBlob)
```

Comment:
any particular reason to not handle the error here?

Reply:
This is because we error out in the upcoming line where we check it the jsonBlob can be safely converted to an object of type map[string]interface{}. If not, we are returning a user-specific error.